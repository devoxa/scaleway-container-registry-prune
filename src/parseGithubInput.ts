const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i
const IMAGE_REGEX = /^rg\.(.*?)\.scw.cloud\/[^/]+\/([^/]+)$/i

export interface GithubInput {
  scwSecretToken: string
  image: string
  tagPattern: string
  keepLast: string
}

export interface Options {
  scwSecretToken: string
  region: string
  imageName: string
  tagPattern: RegExp
  keepLast: number
}

export function parseGithubInput(input: GithubInput): Options {
  if (!input.scwSecretToken.match(UUID_REGEX)) {
    throw new Error('\'scw-secret-token\' is not a valid "secret key" part of a Scaleway API key')
  }

  const imageMatch = input.image.match(IMAGE_REGEX)
  if (!imageMatch) {
    throw new Error("'image' is not the name of an image on the Scaleway Container Registry")
  }
  const region = imageMatch[1]
  const imageName = imageMatch[2]

  let tagPattern
  try {
    tagPattern = new RegExp(input.tagPattern)
  } catch (err) {
    if (!(err instanceof Error)) throw err
    throw new Error("'tag-pattern' is not a valid regular expression")
  }

  if (!input.keepLast.match(/^\d+$/)) {
    throw new Error("'keep-last' is not a valid positive integer")
  }

  const keepLast = parseInt(input.keepLast, 10)
  if (keepLast >= 100) {
    throw new Error("'keep-last' must be smaller than 100")
  }

  return {
    scwSecretToken: input.scwSecretToken,
    region,
    imageName,
    tagPattern,
    keepLast,
  }
}
