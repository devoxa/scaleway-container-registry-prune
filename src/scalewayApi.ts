import querystring from 'querystring'

type ScwGetImagesResponse = { images?: Array<ScwImage> }
type ScwImage = { id: string; name: string }

type ScwGetTagsResponse = { tags?: Array<ScwTag> }
type ScwTag = { id: string; name: string; updated_at: string }

type ScwDeleteTagResponse = { id: string; name: string; updated_at: string }

export async function getImage(
  scwSecretToken: string,
  region: string,
  imageName: string
): Promise<ScwImage> {
  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/images?name=${imageName}`,
    { headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwGetImagesResponse

  if (!json.images || !json.images[0]) {
    throw new Error('Could not get image from the API')
  }

  return json.images[0]
}

export async function listTags(
  scwSecretToken: string,
  region: string,
  imageId: string
): Promise<Array<ScwTag>> {
  const query = querystring.encode({
    page_size: 100,
  })

  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/images/${imageId}/tags?${query}`,
    { headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwGetTagsResponse

  if (!json.tags || json.tags.length === 0) {
    throw new Error('Could not list tags from the API')
  }

  // Sort the tags by the time they were last updated (most recent first)
  const tags = json.tags
  tags.sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime())

  return tags
}

export async function deleteTag(
  scwSecretToken: string,
  region: string,
  tagId: string
): Promise<ScwDeleteTagResponse> {
  const query = querystring.encode({
    force: true,
  })

  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/tags/${tagId}?${query}`,
    { method: 'DELETE', headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwDeleteTagResponse

  if (json.id !== tagId) {
    throw new Error('Could not delete tag from the API')
  }

  return json
}
