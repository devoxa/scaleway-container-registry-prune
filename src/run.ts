import { GithubInput, parseGithubInput } from './parseGithubInput'
import { deleteTag, getImage, listTags } from './scalewayApi'

export async function run(input: GithubInput): Promise<void> {
  const options = parseGithubInput(input)

  const image = await getImage(options.scwSecretToken, options.region, options.imageName)
  const tags = await listTags(options.scwSecretToken, options.region, image.id)

  let count = 0
  for (const tag of tags) {
    if (!tag.name.match(options.tagPattern)) {
      console.log(`Skipping tag: ${tag.name} (${tag.updated_at})`)
      continue
    }

    count++

    if (count <= options.keepLast) {
      console.log(`Keeping tag: ${tag.name} (${tag.updated_at})`)
      continue
    }

    console.log(`Pruning tag: ${tag.name} (${tag.updated_at})`)
    await deleteTag(options.scwSecretToken, options.region, tag.id)
  }
}
