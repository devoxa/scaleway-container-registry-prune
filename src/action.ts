import { getInput } from '@actions/core'
import { run } from './run'

run({
  scwSecretToken: getInput('scw-secret-token'),
  image: getInput('image'),
  tagPattern: getInput('tag-pattern'),
  keepLast: getInput('keep-last'),
}).catch((err) => {
  console.log('Error: ' + err.message)
  process.exit(1)
})
