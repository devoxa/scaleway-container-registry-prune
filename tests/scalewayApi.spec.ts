import { deleteTag, getImage, listTags } from '../src/scalewayApi'

const mockFetch = jest.fn()
global.fetch = mockFetch

function mockFetchJson(value: unknown) {
  mockFetch.mockReturnValue(Promise.resolve({ json: async () => value }))
}

function getFetchCall() {
  return mockFetch.mock.calls[0]
}

describe('scalewayApi', () => {
  beforeEach(() => {
    mockFetch.mockReset()
  })

  describe('getImage', () => {
    test('can get the image from the API', async () => {
      mockFetchJson({
        images: [{ id: 'image-id', name: 'image-name' }],
      })

      const image = await getImage('scw-token', 'scw-region', 'image-name')

      expect(image).toMatchSnapshot()
      expect(getFetchCall()).toMatchSnapshot()
    })

    test('throws an error if the image is not in the API', async () => {
      mockFetchJson({
        images: [],
      })

      let error
      try {
        await getImage('scw-token', 'scw-region', 'image-name')
      } catch (err) {
        error = err
      }

      expect(error).toMatchSnapshot()
    })
  })

  describe('listTags', () => {
    test('can list the tags from the API', async () => {
      mockFetchJson({
        tags: [
          { id: 'tag-id-1', name: 'tag-name-1', updated_at: '2021-01-03' },
          { id: 'tag-id-2', name: 'tag-name-2', updated_at: '2021-01-02' },
          { id: 'tag-id-3', name: 'tag-name-3', updated_at: '2021-01-01' },
        ],
      })

      const tags = await listTags('scw-token', 'scw-region', 'image-id')

      expect(tags).toMatchSnapshot()
      expect(getFetchCall()).toMatchSnapshot()
    })

    test('throws an error if there are no tags in the API', async () => {
      mockFetchJson({
        tags: [],
      })

      let error
      try {
        await listTags('scw-token', 'scw-region', 'image-id')
      } catch (err) {
        error = err
      }

      expect(error).toMatchSnapshot()
    })
  })

  describe('deleteTag', () => {
    test('can delete the tag from the API', async () => {
      mockFetchJson({ id: 'tag-id-1', name: 'tag-name-1', updated_at: '2021-01-03' })

      const tag = await deleteTag('scw-token', 'scw-region', 'tag-id-1')

      expect(tag).toMatchSnapshot()
      expect(getFetchCall()).toMatchSnapshot()
    })

    test('throws an error if the deletion from the API failed', async () => {
      mockFetchJson({ id: 'nope' })

      let error
      try {
        await deleteTag('scw-token', 'scw-region', 'tag-id-1')
      } catch (err) {
        error = err
      }

      expect(error).toMatchSnapshot()
    })
  })
})
