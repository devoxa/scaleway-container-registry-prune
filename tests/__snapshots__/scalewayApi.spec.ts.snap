// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scalewayApi deleteTag can delete the tag from the API 1`] = `
{
  "id": "tag-id-1",
  "name": "tag-name-1",
  "updated_at": "2021-01-03",
}
`;

exports[`scalewayApi deleteTag can delete the tag from the API 2`] = `
[
  "https://api.scaleway.com/registry/v1/regions/scw-region/tags/tag-id-1?force=true",
  {
    "headers": {
      "x-auth-token": "scw-token",
    },
    "method": "DELETE",
  },
]
`;

exports[`scalewayApi deleteTag throws an error if the deletion from the API failed 1`] = `[Error: Could not delete tag from the API]`;

exports[`scalewayApi getImage can get the image from the API 1`] = `
{
  "id": "image-id",
  "name": "image-name",
}
`;

exports[`scalewayApi getImage can get the image from the API 2`] = `
[
  "https://api.scaleway.com/registry/v1/regions/scw-region/images?name=image-name",
  {
    "headers": {
      "x-auth-token": "scw-token",
    },
  },
]
`;

exports[`scalewayApi getImage throws an error if the image is not in the API 1`] = `[Error: Could not get image from the API]`;

exports[`scalewayApi listTags can list the tags from the API 1`] = `
[
  {
    "id": "tag-id-1",
    "name": "tag-name-1",
    "updated_at": "2021-01-03",
  },
  {
    "id": "tag-id-2",
    "name": "tag-name-2",
    "updated_at": "2021-01-02",
  },
  {
    "id": "tag-id-3",
    "name": "tag-name-3",
    "updated_at": "2021-01-01",
  },
]
`;

exports[`scalewayApi listTags can list the tags from the API 2`] = `
[
  "https://api.scaleway.com/registry/v1/regions/scw-region/images/image-id/tags?page_size=100",
  {
    "headers": {
      "x-auth-token": "scw-token",
    },
  },
]
`;

exports[`scalewayApi listTags throws an error if there are no tags in the API 1`] = `[Error: Could not list tags from the API]`;
